### DISCUSSION TITLE
Removing DirichlteBC-dofs to compute eigenvalues

### DISCUSSION THREAD

**Discussion by APaganini** [firedrakeproject/firedrake#2394] [General]
Thanks to slack crowd, I've learnt that to compute eigenvalues of a differential operator it is not sufficient to incorporate `DirichletBC`s by just adding `bcs=bc` in the matrix assembly (e.g. as in `assemble(m, bcs=bc).M.handle`), but that instead one should also remove the dofs associated with the Dirichlet bcs. The demo https://www.firedrakeproject.org/demos/qgbasinmodes.py.html doesn't do it, though. Shall we update it? (Lawrence and Matt have recently ~~learnt~~ taught me how to do that).

_Bonus question:_ In the same demo, PETS options are set to `opts`, but the solver includes them with the spell `es.setFromOptions()`, which does not take opts as an explicit input. Is this correct? (and a sort of PETSc wizardry?)

**Comment by wence-** [comment]
> Thanks to slack crowd, I've learnt that to compute eigenvalues of a differential operator it is not sufficient to incorporate `DirichletBC`s by just adding `bcs=bc` in the matrix assembly (e.g. as in `assemble(m, bcs=bc).M.handle`), but that instead one should also remove the dofs associated with the Dirichlet bcs. The demo https://www.firedrakeproject.org/demos/qgbasinmodes.py.html doesn't do it, though. Shall we update it? (Lawrence and Matt have recently ~~learnt~~ taught me how to do that).

Please go ahead and do so.

> _Bonus question:_ In the same demo, PETS options are set to `opts`, but the solver includes them with the spell `es.setFromOptions()`, which does not take opts as an explicit input. Is this correct? (and a sort of PETSc wizardry?)

The `opts` reference the _global_ PETSc options database. `setFromOptions()` inspects the options database to configure things.

**Reply by APaganini** [reply]
1. I'll write the update
2. Many thanks for claryfing that it is processed via global variables
3. Apologies for a typo. I originally wrote _Lawrence and Matt have recently_ **learnt** _how to do that_ whereas I obviously meant _Lawrence and Matt have recently_ **taught me** _how to do that_ :)

**Reply by APaganini** [reply]
The update is here: #2402
